Sources/SwiftSyntaxMacroExpansion/MacroSystem.swift

```diff
@@ -1204,10 +1218,33 @@ extension MacroApplication {
 
 // MARK: Freestanding macro expansion
 
+protocol MacroExpansion<ResultType> {
+  associatedtype ResultType
+  func map<T>(_ transform: (ResultType) throws -> T) rethrows -> T
+}
+
 extension MacroApplication {
+  struct StackMacroExpansion<ResultType>: MacroExpansion {
+    let expandedNode: ResultType
+    unowned let macroApplication: MacroApplication
+
+    init(expandedNode: ResultType, macro: any Macro.Type, macroApplication: MacroApplication) {
+      self.expandedNode = expandedNode
+      self.macroApplication = macroApplication
+      macroApplication.expandingFreestandingMacros.append(macro)
+    }
+
+    func map<T>(_ transform: (ResultType) throws -> T) rethrows -> T {
+      defer {
```

Comment:
Would it be possible to run `macroApplication.expandingFreestandingMacros.append(macro)` right before `transform` instead of in `init`? That way, we’d guarantee that the append is always balanced with a `remove` and they can never get out-of-sync if somebody forgets to calls `map`.

Reply:
True, I was carried away by the thought of storing one less field in the struct...